### PR TITLE
Implement Hessian support to JacobianTape

### PR DESCRIPTION
**Pull request by soudy**
**Context:**
The gradient of quantum tapes can be computed via the parameter-shift rule. The Hessian of quantum tapes can be calculated similarly as shown in https://arxiv.org/abs/2008.06517 (eqns 11-13). 

**Description of the Change:**
This adds the `QubitParamShift.hessian` method for calculating the second derivative with respect to tape parameters `i` and `j` using the second-order shift formula. It also adds the `JacobianTape.hessian` function for computing the Hessian with respect to all trainable parameters on the quantum tape.

**Benefits:**
Adding support for second-order derivatives of quantum circuits allows for the use of second-order optimization methods like the Newton optimizer and the quantum natural gradient optimizer. 

**Possible Drawbacks:**
None

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/948
https://github.com/PennyLaneAI/pennylane/issues/890

### CODE REVIEW

File: tests/tape/tapes/test_qubit_param_shift.py [hunk 1]
```diff
@@ -541,3 +541,38 @@ def test_expval_and_variance(self, tol):
...
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        hessian = tape.hessian(dev, method="analytic")
+        expected = np.array([
+            [ 0.794,  0.055,  0.109, -0.145,  0.   ],
+            [ 0.055,  0.794, -0.042,  0.056, -0.   ],
+            [ 0.109, -0.042,  0.794,  0.11 ,  0.   ],
+            [-0.145,  0.056,  0.11 ,  0.794,  0.   ],
+            [ 0.   , -0.   ,  0.   ,  0.   , -0.   ]
+        ])
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/tape/tapes/test_qubit_param_shift.py:576]: I ran this test locally, and separately computed the symbolic hessian for this circuit, and can confirm this is correct 🎉

File: tests/tape/tapes/test_qubit_param_shift.py [hunk 2]
```diff
@@ -541,3 +541,38 @@ def test_expval_and_variance(self, tol):
...
+        with QubitParamShiftTape() as tape:
+            for i in range(n_wires):
+                qml.RX(weights[i], wires=i)
+
+            qml.CNOT(wires=[0, 1])
+            qml.CNOT(wires=[2, 1])
+            qml.CNOT(wires=[3, 1])
+            qml.CNOT(wires=[4, 3])
+
+            qml.expval(qml.PauliZ(1))
```
Comment (josh146) [inline tests/tape/tapes/test_qubit_param_shift.py:563]: It might also be good to have a test that tests the Hessian computation for a vector-valued tape! e.g., one that has multiple expval returns, or a probability return.

**Review by josh146 — COMMENTED** [review]
Thanks for this great contribution @soudy! I left some general comments and suggestions. In particular, it would be nice to reduce the number of executions required to compute the hessian using some tips and tricks in the original paper.

File: pennylane/tape/tapes/jacobian_tape.py [hunk 3]
```diff
@@ -546,3 +549,84 @@ def jacobian(self, device, params=None, **options):
...
         return jac
+
+    def hessian(self, device, params=None, **options):
+        if any([m.return_type is State for m in self.measurements]):
+            raise ValueError("The Hessian method does not support circuits that return the state")
+
+        method = options.get("method", "analytic")
+
+        if method != "analytic":
+            raise ValueError(f"Unknown Hessian method '{method}'")
```
[the remark this answers is not in the document]
Comment (soudy) [reply pennylane/tape/tapes/jacobian_tape.py:560]: Yeah this is what I noticed too. With some optimizations added to the `JacobianTape.hessian` function, they are becoming quite different functions structure-wise. I've tried to factor out some of the common logic into utility functions, but let me know if you see more opportunity to reduce duplication.

File: pennylane/tape/tapes/qubit_param_shift.py [hunk 4]
```diff
@@ -271,3 +273,64 @@ def processing_fn(results):
...
+            param_shift = op.get_parameter_shift(p_idx, shift=s)
+            param_shifts.append(param_shift)
+
+        coeffs = []
+        tapes = []
+        shift = np.eye(len(params))
+
+        for (_, _, s1) in param_shifts[0]:
+            for (_, _, s2) in param_shifts[1]:
+                c = 0.25 / np.sin(s1) / np.sin(s2)
```
[the remark this answers is not in the document]
Comment (soudy) [reply pennylane/tape/tapes/qubit_param_shift.py]: Woops, brainfart. Good catch.

File: tests/tape/tapes/test_qubit_param_shift.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (soudy) [reply tests/tape/tapes/test_qubit_param_shift.py]: I took this example from the https://arxiv.org/pdf/2008.06517.pdf paper's numerical results.

File: pennylane/tape/tapes/qubit_param_shift.py [hunk 5]
```diff
@@ -271,3 +273,64 @@ def processing_fn(results):
...
+            s = options.get(shift_kwarg, np.pi / 2)
+            param_shift = op.get_parameter_shift(p_idx, shift=s)
+            param_shifts.append(param_shift)
+
+        coeffs = []
+        tapes = []
+        shift = np.eye(len(params))
+
+        for (_, _, s1) in param_shifts[0]:
+            for (_, _, s2) in param_shifts[1]:
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/tape/tapes/qubit_param_shift.py]: It might be the case that re-using a previous gradient/jacobian computation is too complicated, and is left to a future PR

**Comment by soudy** [comment]
Thanks for the input @josh146. I've implemented your feedback, please let me know what you think. I would like to add more tests, but I'll have to take some more time for that somewhere.

File: pennylane/tape/tapes/jacobian_tape.py [hunk 6]
```diff
@@ -546,3 +569,163 @@ def jacobian(self, device, params=None, **options):
...
+        the tapes' parameters in-place:
+
+        >>> tape.get_parameters()
+        [2.73943676, 0.16289932, 3.4536312, 2.73521126, 2.6412488]
+
+        If a tape has no trainable parameters, the Hessian will be empty:
+
+        >>> tape.trainable_params = {}
+        >>> tape.hessian(dev)
+        array([], shape=(1, 0), dtype=float64)
```
Comment (soudy) [inline pennylane/tape/tapes/jacobian_tape.py:646]: What should we return when there are no trainable parameters? Right now I return a `array([], shape=(1, 0), dtype=float64)`, but I'm not sure what the best practice is here.

File: pennylane/tape/tapes/qubit_param_shift.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (soudy) [reply pennylane/tape/tapes/qubit_param_shift.py]: I've left re-using a previous jacobian for a future PR, but I did add the optimization for the `s1 = s2 = np.pi/2` and `s1 = s2 = pi/4` cases.

**Comment by soudy** [comment]
Also, regarding the CodeFactor check, would you like me to make functions out of `_has_trainable_params` and `_flatten_processing_result` instead of class methods? If so, where should I put them?

**Comment by josh146** [comment]
> Also, regarding the CodeFactor check, would you like me to make functions out of _has_trainable_params and _flatten_processing_result instead of class methods? If so, where should I put them?

In this case, it's fine for them to be static methods! This way they are namespaced within the class that they are used, and easily accessible by subclasses when needed (while avoiding circular imports).

File: pennylane/tape/tapes/jacobian_tape.py (hunk 6 above)
Comment (josh146) [reply pennylane/tape/tapes/jacobian_tape.py:647]: That's a good question. We should match what the `jacobian()` method does, but I just had a look, and suspect that it might be ambiguous (sometimes it might return `None`, sometimes a zeros array of size `(qnode_output, params)`).

**Comment by soudy** [comment]
@josh146 Gotcha. This PR is ready for a final review I think.

File: pennylane/tape/tapes/jacobian_tape.py (hunk 6 above)
Comment (josh146) [reply pennylane/tape/tapes/jacobian_tape.py:649]: Reminder to myself that we should double check that the `jacobian()` method (and the `hessian` method) have consistent output for no trainable parameters

**Comment by josh146** [comment]
> @josh146 Gotcha. This PR is ready for a final review I think.

Thanks @soudy! I will try and have a look over the next few days

File: pennylane/tape/tapes/qubit_param_shift.py [hunk 7]
```diff
@@ -271,3 +272,85 @@ def processing_fn(results):
...
+        else:
+            # no optimizations can be made, generate all 4 tapes
+            for idx in idxs:
+                t_idx = list(self.trainable_params)[idx]
+                op = self._par_info[t_idx]["op"]
+                p_idx = self._par_info[t_idx]["p_idx"]
+                s = idx_shifts[idx]
+
+                param_shift = op.get_parameter_shift(p_idx, shift=s)
+                param_shifts.append(param_shift)
```
Comment (josh146) [inline pennylane/tape/tapes/qubit_param_shift.py:331]: I _think_ this for loop could be moved to line 309 as an `else` branch:

```python
if i == j and idx_shifts[i] == idx_shifts[j]:
    # optimize the number of parameter shifts
    if idx_shifts[i] == np.pi / 2:
        ...
    elif idx_shifts[i] == np.pi / 4:
        ...
else:
    # general 4-term second-derivative shift
    for idx in idxs:
        t_idx = list(self.trainable_params)[idx]
        op = self._par_info[t_idx]["op"]
        p_idx = self._par_info[t_idx]["p_idx"]
        s = idx_shifts[idx]

        param_shift = op.get_parameter_shift(p_idx, shift=s)
        param_shifts.append(param_shift)
```

Comment (josh146) [reply pennylane/tape/tapes/qubit_param_shift.py:331]: This has the (potential) advantage that the loop below generating the shifts can be used for any form of the `param_shifts` list. Although I might be missing something.

File: tests/tape/tapes/test_qubit_param_shift.py [hunk 8]
```diff
@@ -541,3 +542,55 @@ def test_expval_and_variance(self, tol):
...
         assert gradF == pytest.approx(expected, abs=tol)
+
+class TestHessian:
+    """Tests for parameter Hessian method"""
+
+    @pytest.mark.parametrize("s1", [np.pi / 2, np.pi / 4, 2])
+    @pytest.mark.parametrize("s2", [np.pi / 2, np.pi / 4, 3])
+    @pytest.mark.parametrize("G", [qml.RX, qml.RY, qml.RZ, qml.PhaseShift])
+    def test_pauli_rotation_hessian(self, s1, s2, G, tol):
+        """Tests that the automatic Hessian of Pauli rotations are correct."""
```
Comment (josh146) [inline tests/tape/tapes/test_qubit_param_shift.py:553]: Good idea adding this test!

**Review by josh146 — COMMENTED** [review]
@soudy thanks for putting in the extra work to tidy this up! Looking almost ready, left a few more comments.

File: pennylane/tape/tapes/jacobian_tape.py [hunk 9]
```diff
@@ -546,3 +571,157 @@ def jacobian(self, device, params=None, **options):
...
+
+        params = np.array(params)
+
+        # perform gradient method validation
+        diff_methods = self._grad_method_validation(method)
+
+        if not self._has_trainable_params(params, diff_methods):
+            # Either all parameters have grad method 0, or there are no trainable
+            # parameters. Simply return an empty Hessian.
+            return np.zeros((self.output_dim, len(params)), dtype=float)
```
[the remark this answers is not in the document]
Comment (soudy) [reply pennylane/tape/tapes/jacobian_tape.py:670]: Good catch, missed this one.

File: pennylane/tape/tapes/jacobian_tape.py [hunk 10]
```diff
@@ -546,3 +571,157 @@ def jacobian(self, device, params=None, **options):
...
+            # extract the correct results from the flat list
+            res = results[start : start + res_len]
+            start += res_len
+
+            # postprocess results to compute the gradient
+            g = self._flatten_processing_result(processing_fn(res))
+
+            if hessian is None:
+                # update the tape's output dimension
+                self._output_dim = len(params)
```
[the remark this answers is not in the document]
Comment (soudy) [reply pennylane/tape/tapes/jacobian_tape.py:718]: You're right, wasn't sure this was being used somewhere else but seems like it isn't.

File: pennylane/tape/tapes/qubit_param_shift.py (hunk 7 above)
Comment (soudy) [reply pennylane/tape/tapes/qubit_param_shift.py:331]: I thought of this, but I think this fails when `i == j and idx_shifts[i] == idx_shifts[j]` is true, but `idx_shifts[i]` is not `np.pi / 2` or `np.pi / 4`. That would require another `else` branch inside the first `if` which would do the same thing as the outer else branch.

**Comment by soudy** [comment]
Thanks @josh146, applied your feedback. Let me know if I missed something.

File: pennylane/tape/tapes/jacobian_tape.py [hunk 11]
```diff
@@ -546,3 +577,155 @@ def jacobian(self, device, params=None, **options):
...
+            device (.Device, .QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+
+        Keyword Args:
+            method="analytic" (str): The differentiation method. Currently only
+                supports ``"analytic"``.
+            s1=pi/2 (float): the size of the shift for index i in the parameter-shift Hessian computations
+            s2=pi/2 (float): the size of the shift for index j in the parameter-shift Hessian computations
```
Comment (albi3ro) [inline pennylane/tape/tapes/jacobian_tape.py:603]: Just an idea:
I'm assuming most use cases will have the same values for `s1` and `s2`.  
We could allow the user to either: 
* supply one number `s` that applies to both 
* supply `s1` and `s2` individually

**Review by albi3ro — COMMENTED** [review]
Good work 👍 

I left one comment on being able to pass one shift that applies to both variables.  

Other than that, I can't find anything to say other than "looks good!"

File: pennylane/tape/tapes/jacobian_tape.py (hunk 11 above)
Comment (josh146) [reply pennylane/tape/tapes/jacobian_tape.py:603]: Alternatively, `s` could be a float _or_ a list!

```python
hessian(params, s=[pi/2, pi/4])
hessian(params, s=pi/2)
```

Although leaving it as two specific arguments is more explicit, and probably better for an internal method.

File: pennylane/tape/tapes/qubit_param_shift.py (hunk 7 above)
Comment (josh146) [reply pennylane/tape/tapes/qubit_param_shift.py:334]: Got it!

**Review by josh146 — APPROVED** [review]
Thanks for finishing this off @soudy, this is a great addition to PennyLane! From my end this is now ready to be merged, and then we can work on integrating this low-level method with TensorFlow, PyTorch, and Autograd :)

File: pennylane/tape/tapes/jacobian_tape.py (hunk 11 above)
Comment (soudy) [reply pennylane/tape/tapes/jacobian_tape.py:603]: Good idea! I do like the idea of adding the possibility to define both shifts through one parameter. Let me know if you guys want me to implement this, and which method, or if you want me to leave it as is for now.

Comment (co9olguy) [reply pennylane/tape/tapes/jacobian_tape.py:603]: Thanks @soudy. No worries about implementing this idea now, the PR is already a great contribution as-is

**Comment by josh146** [comment]
@soudy I believe this is now ready to be merged in 🙂 Let me know if that is not the case, or if there is any other changes you plan on including!

**Comment by soudy** [comment]
@josh146 Ready to merge from my side!

**Comment by co9olguy** [comment]
Awesome, thanks for the solid contribution @soudy!

**Comment by cvjjm** [comment]
Great that this has been merged!

One comment: Looing at the implementation I get the impression that the hessian computation will still only work for circuits whose parametrized gates are differentiable with a two-shift rule (or am I missing something?).

Are you planning to generalize this to four shift rule gates?

**Comment by josh146** [comment]
Hi @cvjjm, yes you are correct - at the moment, only the two-shift rule is supported on hardware if using the parameter-shift rule. If using backprop, however, all gates should be supported.

However, this is actively on our roadmap, and we are investigating how to extend this to _any_ shift rule, as well as nth order derivatives!
